src/freertos_drivers/tinyusb/TinyUsbCdcImpl.hxx

```diff
@@ -0,0 +1,436 @@
+                 * critical section lock.
+                 */
+                DeviceBufferBase::block_until_condition(file, false);
+            }
+        }
+
+        count -= bytes_written;
+        result += bytes_written;
+        data += bytes_written;
+    }
+
+    if (!result && (file->flags & O_NONBLOCK))
+    {
+        return -EAGAIN;
+    }
+
+    if (result)
+    {
+        tud_cdc_write_flush();
+    }
+
+    return result;
+}
+
+/** Device select method. Default impementation returns true.
+ * @param file reference to the file
+ * @param mode FREAD for read active, FWRITE for write active, 0 for
+ *        exceptions
+ * @return true if active, false if inactive
+ */
+bool TinyUsbCdc::select(File *file, int mode)
+{
+    bool retval = false;
+    switch (mode)
+    {
+        case FREAD:
+            portENTER_CRITICAL();
+            if (tud_cdc_available() > 0)
+            {
+                retval = true;
+            }
+            else
+            {
+                Device::select_insert(&selectInfoRead_);
+            }
+            portEXIT_CRITICAL();
+            break;
+        case FWRITE:
+            portENTER_CRITICAL();
+            if (tud_cdc_write_available() > 0)
+            {
+                retval = true;
+            }
+            else
+            {
+                Device::select_insert(&selectInfoWrite_);
+            }
+            portEXIT_CRITICAL();
+            break;
+        default:
+        case 0:
+            /* we don't support any exceptions */
+            break;
+    }
+    return retval;
+}
+
+inline void TinyUsbCdc::rx_available() {
+    Device::select_wakeup(&selectInfoRead_);
+}
+
+inline void TinyUsbCdc::tx_complete() {
+    Device::select_wakeup(&selectInfoWrite_);
+}
+
+extern "C" {
+
+// Invoked when CDC interface received data from host
+void tud_cdc_rx_cb(uint8_t itf)
+{
+  (void) itf;
+  Singleton<TinyUsbCdc>::instance()->rx_available();
+}
+
+// Invoked when a TX is complete and therefore space becomes available in TX buffer
+void tud_cdc_tx_complete_cb(uint8_t itf)
+{
+  (void) itf;
+  Singleton<TinyUsbCdc>::instance()->tx_complete();
+}
+
+
+// ===================== USB DESCRIPTORS =============================
+
+/* A combination of interfaces must have a unique product id, since PC will save device driver after the first plug.
+ * Same VID/PID with different interface e.g MSC (first), then CDC (later) will possibly cause system error on PC.
+ *
+ * Auto ProductID layout's Bitmap:
+ *   [MSB]         HID | MSC | CDC          [LSB]
+ */
+#define _PID_MAP(itf, n)  ( (CFG_TUD_##itf) << (n) )
+#define USB_PID           (0x4000 | _PID_MAP(CDC, 0) | _PID_MAP(MSC, 1) | _PID_MAP(HID, 2) | \
+                           _PID_MAP(MIDI, 3) | _PID_MAP(VENDOR, 4) )
+
+#define USB_VID   0xCafe
+#define USB_BCD   0x0200
+
+// Invoked when received GET DEVICE DESCRIPTOR
+// Application return pointer to descriptor
+uint8_t const *tud_descriptor_device_cb(void) {
```

Comment:
Incorrect open brace style, here and below.

Reply:
fixed all whitespace